src/icon-api.js

```diff
@@ -0,0 +1,144 @@
+  #dataStore
+
+  /**
+   * @param {Object} opts
+   * @param {import('./datatype/index.js').DataType<
+   *   import('./datastore/index.js').DataStore<'config'>,
+   *   typeof import('./schema/project.js').iconTable,
+   *   'icon',
+   *   import('@mapeo/schema').Icon,
+   *   import('@mapeo/schema').IconValue
+   * >} opts.iconDataType
+   * @param {import('./datastore/index.js').DataStore<'config'>} opts.iconDataStore
+   * @param {string} [opts.projectId]
+   */
+  constructor({ iconDataType, iconDataStore, projectId }) {
+    this.#dataType = iconDataType
+    this.#dataStore = iconDataStore
+    this.#projectId = projectId
+  }
+
+  /**
+   * @param {Omit<import('@mapeo/schema').IconValue, 'variants'> &
+   * {variants: Array<
+   *    import('@mapeo/schema/dist/proto/icon/v1.js').Icon_1_IconVariant
+   *    & {blob: Buffer}>}} icon
+   */
+  async [kCreate](icon) {
+    if (icon.variants.length < 1) {
+      throw new Error('empty variants array')
+    }
+    const blobsIds = await Promise.all(
+      icon.variants.map(({ blob }) => {
+        return this.#dataStore.writeRaw(blob)
+      })
+    )
+    const variantsWithId = icon.variants.map((variant, i) => {
+      // eslint-disable-next-line no-unused-vars
+      const { blob, ...variantWithoutBlob } = variant
+      return { ...variantWithoutBlob, blobVersionId: blobsIds[i] }
+    })
+    return await this.#dataType.create({
+      schemaName: 'icon',
+      name: icon.name,
+      // @ts-ignore TODO: remove
+      variants: variantsWithId,
+    })
+  }
+
+  /** @typedef {'image/png' | 'image/svg+xml'} ValidMimeType */
+
+  /**
+   * @param {Object} opts
+   * @param {String} opts.iconId
+   * @param {String} [opts.size]
+   * @param {number} [opts.pixelDensity]
+   * @param {ValidMimeType} [opts.mimeType]
+   */
+  async [kGetIcon]({ iconId, size, pixelDensity, mimeType }) {
+    const iconRecord = await this.#dataType.getByDocId(iconId)
+    const iconVariant = this[kGetBestVariant](iconRecord.variants, {
+      size,
+      pixelDensity,
+      mimeType,
+    })
+    const blob = await this.#dataStore.readRaw(iconVariant.blobVersionId)
+    return { icon: blob, mimeType: iconVariant.mimeType }
+  }
+
+  /**
+   * @param {Object} opts
+   * @param {String} opts.iconId
+   * @param {String} [opts.size]
+   * @param {number} [opts.pixelDensity]
+   * @param {ValidMimeType} [opts.mimeType]
+   */
+  async [kGetIconUrl]({ iconId, size, pixelDensity }) {
+    return `/${this.#projectId}/${iconId}/${size}/${pixelDensity}`
+  }
+
+  /** @typedef {{
+   * mimeType: ValidMimeType,
+   * size: string,
+   * pixelDensity: number,
+   * score: number
+   * }} IconVariant */
+
+  /**
+   * @param {import('@mapeo/schema').IconValue['variants']} variants
+   * @param {object} opts
+   * @param {string} [opts.size]
+   * @param {number} [opts.pixelDensity]
+   * @param {ValidMimeType} [opts.mimeType]
+   **/
+  [kGetBestVariant](variants, { size, pixelDensity, mimeType = 'image/png' }) {
```

Comment:
Yeah, so my confusion I think spans for not knowing what parameters we have control of. 
The implementation here basically 'scores'  each variant given matching `pixelDensity` or `size`; so a variant with:
* same `pixelDensity` -> score of 1
* same `pixelDensity` and `size` -> score of 2
From that we choose the variant with best score and if there's a tie (two variants with same score) we choose the first one in the array.
So, given you comment I think we can simplify it by expecting a matching size and throwing if it doesn't exist (same as we do with `mimeType`).
From that, given we only have one criteria (`pixelDensity`) we can basically choose the variant with matching or closest `pixelDensity`. 
Also we eliminate the posibility of _ties_

Am I making sense??

Reply:
> Yeah, so my confusion I think spans for not knowing what parameters we have control of.

To be clear, I'm _also_ not sure 😄 My initial comment really stems from me not really knowing real-world cases that would benefit from having a fallback behavior for this, as opposed to just erroring completely.

Also just realized that my reference to MapboxGL doesn't really apply in this case because these icons aren't used on the actual map for Mapeo (they're used for categories). Still a good example of understanding which parameters may not be in our control, I guess...

@gmaclennan if you have any additions to this, very much welcomed

> So, given you comment I think we can simplify it by expecting a matching size and throwing if it doesn't exist (same as we do with mimeType).
> From that, given we only have one criteria (pixelDensity) we can basically choose the variant with matching or closest pixelDensity.
> Also we eliminate the posibility of ties

if my assumption about the `size` and `pixelDensity` parameters are okay, then yeah I _think_ this makes sense.

Reply:
Do you think @achou11 that moving `getBestVariant` to `src/utils.js` makes sense??

Reply:
No strong opinion but I don' think it's strictly necessary if it's not being used elsewhere. I think exporting as a separate variable from the IconApi file would be fine

Reply:
- No matching mimetype -> throw notfound
- No matching pixeldensity -> next lowest pixel density -> no lower pd -> next higher pd
- No matching size -> next smaller size -> no smaller size -> next higher size
- Prefer matching size over matching pd

The sizes for icons are relevant for bitmap icons (e.g. PNG), for SVG they are only relevant if the icon creator has chosen to make a different design of icon for display at smaller sizes - e.g. simplified

Reply:
@gmaclennan should all of those parameters be required or not? If not, wondering how the priority changes, or would we set defaults?

Currently the implementation marks mime type, size, and pixel density as optional. Would think that maybe mime type should be required at the very least? If I don't specify size, which size takes priority in the case of many valid matches? Same question for pixel density...

e.g.

```ts
// Assume these two are basically the same except for their size and blobVersionId
const smallVariant = {...}
const mediumVariant = { ... }

// What's the result here?
const result = getBestVariant([smallVariant, mediumVariant], { mimeType: ..., pixelDensity: ... })
```